spec/forms/time_locking_periods/update_form_spec.rb

```diff
@@ -0,0 +1,122 @@
+        time_record_2 = create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 2.days, description: nil)
+        time_record_3 = create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 3.days, project_id: nil)
+        form = build_form({ approved: true }, period)
+        form.valid?
+        dates = [time_record_1.assigned_date, time_record_3.assigned_date].join(", ")
+        expect(form.errors[:base]).to include(I18n.t("time_locking_periods.has_inconsistent_data"))
+      end
+
+      it 'should add dates to the form\'s variable' do
+        time_record_1 = create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 2.days, description: nil)
+        time_record_2 = create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 2.days, description: nil)
+        time_record_3 = create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 3.days, project_id: nil)
+        form = build_form({ approved: true }, period)
+        form.valid?
+        dates = [time_record_1.assigned_date, time_record_3.assigned_date]
+        expect(form.dates_of_invalid_time_records).to eq(dates)
+      end
+
+      it 'should not add error in case of having only consisent data' do
+        create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today + 2.days, description: nil)
+        create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 2.days)
+        create(:time_record, workspace: period.workspace, user: period.user, assigned_date: Date.today - 3.days)
+        form = build_form({ approved: true }, period)
+        form.valid?
+        expect(form.errors[:base]).to be_empty
+      end
+    end
+
+  end
+
+  describe ".initialize" do
+
+    it "should add period to the form" do
+      form = build_form({}, period)
+      expect(form.period).to eq(period)
+    end
+
+    it "should pass attributes to the form" do
+      form = build_form({ approved: true }, period)
+      expect(form.approved).to be_truthy
+    end
+  end
+
+  describe '#persist!' do
+    let(:form) { build_form({ approved: true }, period) }
+
+    it 'should update real period with passed data' do
+      form.persist!
+      expect(period.reload.approved).to be_truthy
+    end
+
+    it "should change telegram message text if period has that" do
+      allow(period).to receive(:telegram_message_id) { 133 }
+      allow(period.user).to receive(:telegram_id) { 122 }
+      text = "new message"
+      allow(I18n).to receive(:t).with(
+        'telegram.period_was_succesfully_approved',
+        workspace: period.workspace.name,
+        from: period.beginning_of_period,
+        to: period.end_of_period
+      ) { text }
+      expect(Telegram.bot).to receive(:edit_message_text).with(
```

Comment:
Is it OK to call expect before tested code?

Reply:
what do you mean? I've set the expectation about receiving the method. Is it possible to do this after the launching code?

Reply:
I mean in previous test (line 81) `expect` called after `form.persist!` and in this test it's opposite - `form.persist!` called after `expect`. Maybe you can point me out what to read about it because I'm not sure how this `expect` works.
I expect it to run after all methods fired like any other ruby code.